### ISSUE TITLE
Applying clippy suggestion leads to inference error (`or_insert_with` / `or_default`)

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-clippy#11383]
### Summary

When using `map.entry(e).or_insert_with(Type::new)` where `Type` has an identical `Default` implementation, clippy suggests to use `.or_default()` even if the value type of the map hasn't been declared, so using something like `Type::new` is required for inference.

### Lint Name

unwrap_or_default

### Reproducer

I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2c948689484b4cd957f24e1dd559e976)):

```rust
use std::collections::{BTreeMap, BTreeSet};

fn main() {
    let arr = [("a", 1), ("b", 2), ("a", 3)];
    let map = arr.iter().fold(BTreeMap::new(), |mut acc, (k, v)| {
        acc.entry(k).or_insert_with(BTreeSet::new).insert(v);
        acc
    });
    println!("{map:?}");
}
```

I saw this happen:

```
warning: use of `or_insert_with` to construct default value
 --> src/main.rs:6:22
  |
6 |         acc.entry(k).or_insert_with(BTreeSet::new).insert(v);
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `or_default()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unwrap_or_default
  = note: `#[warn(clippy::unwrap_or_default)]` on by default
```

I expected to see this happen: No warning, since there isn't really a better way to write this.

### Version

```text
rustc 1.74.0-nightly (ef85656a1 2023-08-21)
binary: rustc
commit-hash: ef85656a10657ba5e4f7fe2931a4ca6293138d51
commit-date: 2023-08-21
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
